src/driver/kvm.rs

```diff
@@ -474,6 +455,1014 @@ mod tests {
+
+        let vcpu: u16 = 0;
+        let intercept_type = InterceptType::Breakpoint;
+        let enabled = true;
+        let result = Kvm::toggle_intercept(&mut kvm, vcpu, intercept_type, enabled);
+
+        assert!(result.is_err(), "Expected error, got ok instead!");
+    }
+
+    #[test]
+    fn test_pf_intercept_succeeds() {
+        let mut kvmi_mock = MockKVMi::default();
+        kvmi_mock
+            .expect_control_cr()
+            .times(0)
+            .returning(|_, _, _| Ok(()));
+        kvmi_mock
+            .expect_control_msr()
+            .times(0)
+            .returning(|_, _, _| Ok(()));
+        kvmi_mock
+            .expect_control_events()
+            .times(0)
+            .returning(|_, _, _| Ok(()));
+        kvmi_mock
+            .expect_control_events()
+            .with(
+                eq(0),
+                function(|x| matches!(x, KVMiInterceptType::Pagefault)),
+                eq(true),
+            )
+            .times(1)
+            .returning(|_, _, _| Ok(()));
+        setup_kvmi_mock(&mut kvmi_mock);
+        let mut kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+
+        let vcpu: u16 = 0;
+        let intercept_type = InterceptType::Pagefault;
+        let enabled = true;
+        let result = Kvm::toggle_intercept(&mut kvm, vcpu, intercept_type, enabled);
+
+        assert!(result.is_ok(), "Expected ok, got error instead!");
+    }
+
+    #[test]
+    fn test_pf_intercept_fails_if_kvmi_conrol_events_fails() {
+        let mut kvmi_mock = MockKVMi::default();
+        kvmi_mock.expect_init().times(1).returning(|_| Ok(()));
+        kvmi_mock.expect_get_vcpu_count().returning(move || Ok(1));
+        kvmi_mock
+            .expect_control_events()
+            .with(
+                eq(0),
+                function(|x| matches!(x, KVMiInterceptType::Cr)),
+                eq(true),
+            )
+            .times(1)
+            .returning(|_, _, _| Ok(()));
+        kvmi_mock
+            .expect_control_events()
+            .with(
+                eq(0),
+                function(|x| matches!(x, KVMiInterceptType::Cr)),
+                eq(false),
+            )
+            .times(1)
+            .returning(|_, _, _| Ok(()));
+        kvmi_mock
+            .expect_control_events()
+            .with(
+                eq(0),
+                function(|x| matches!(x, KVMiInterceptType::Msr)),
+                eq(true),
+            )
+            .times(1)
+            .returning(|_, _, _| Ok(()));
+        kvmi_mock
+            .expect_control_events()
+            .with(
+                eq(0),
+                function(|x| matches!(x, KVMiInterceptType::Msr)),
+                eq(false),
+            )
+            .times(1)
+            .returning(|_, _, _| Ok(()));
+        kvmi_mock
```

Comment:
@Wenzel is it logically possible for toggling pagefault interception on to fail if called a second time?